Improve `array_on_device` performance via subloc-aware comm calls

e-kayrakli opened the pull request:
This PR's main purpose is to improve local, but wide pointer based data movements' performance. The most obvious case where that matters is array initialization on CPU.

Resolves https://github.com/Cray/chapel-private/issues/4571
Resolves https://github.com/Cray/chapel-private/issues/4596

### The main change: subloc-aware comm
- Adds `chpl_gen_comm_get_from_subloc` and `chpl_gen_comm_put_to_subloc`
  - These take an additional argument for the sublocale of the "remote" data.
  - They generate the sublocale of the "local" data via `chpl_task_getRequestedSubloc()`
  - If both source and the destination are within this locale, they'll both call `chpl_gpu_memcpy`
  - If they aren't, they will call the new interface `chpl_gpu_comm_get` and `chpl_gpu_comm_put`
- Renames `chpl_gpu_memmove` to `chpl_gpu_memcpy` and improves its implementation
  - With the sublocale information we can now fast-track all-host data movement. See "Caveat" below for more.
- Rewires codegen to generate and pass sublocale to the new comm functions in the runtime.
- Generates and conveys sublocale information in `DefaultRectangular` where bulk array comm is implemented.
- Adds data movement diagnostics to `GpuDiagnostics`
- Adds a performance and a correctness test
- Bunch of test changes to make them more suitable for execution with `array_on_device`:
  - With this PR, we need to use smaller `on` statements if we want decent performance with array_on_device. e.g. using `for` inside an `on` statement targeting a GPU results in a CPU iteration with fine grained access, which kills performance. 
  - Refactors the helper function to verify gpu diagnostics into a new common testing module. That's an improvement, but as a very near term todo, we want to move that function to `GpuDiagnostics` directly. See https://github.com/Cray/chapel-private/issues/5135


#### Caveat
When we have a wide pointer whose sublocale is -2, it signals that the data was created on host. This is enough guarantee that the allocation for it was made on host memory. However, the inverse is not true. If you have a sublocale 0 on a wide pointer, it only tells you that the variable was created on `on here.gpus[0]`. However, it could be an `int` that's not subject to GPU allocation. To be able to handle that, we only fast-track host-only data movement. For any sublocale that's not -2, we still ask the implementation layer for whereabouts of the data.

A potential incremental improvement is to handle cases where we can tell at least one of the source/destination to be on host memory. https://github.com/Cray/chapel-private/issues/5152 has more information on this.


### While-there changes
- Moves context handling to the `chpl-gpu` interface from `chpl-gpu-impl` interface
- Adds a proper `chpl_gpu_impl_mem_array_alloc` that's called from `chpl_gpu_mem_array_alloc`.
- Removes the `chpl_gpu_device_alloc` workaround
  - We can now remove it because of (1) proper context handling as we were not setting context properly in one case, (2) we have proper get/put implementation where we can actually put the locale instance in GPU memory and get/put after 
- Removes redundant global setting for `chpl_nodeID`


#### Test status
- [x] nvidia + `array_on_device`
- [x] amd + `array_on_device`
- [x] nvidia + `unified_memory`
- [x] amd + `unified_memory`
- [x] nvidia + gasnet + `array_on_device`
- [x] cpu
- [x] standard
- [x] gasnet

e-kayrakli commented:
With this branch, array_on_device performs better than unified_memory across the board. Here are some results from SHOC suite. I haven't done much interpretation of magnitudes just yet.

SHOC Benchmark | Improvement (x)
-- | --
Sort | 1.36
Sort (Kernel) | 1.20
triad 64K | 1.95
triad 16M | 2.02
triad 64K (kernel) | 5.73
triad 16M (kernel) | 14.00
triadalt1 64K | 2.93
triadalt1 16M | 2.16
triadalt1 64K (kernel) | 6.07
triadalt1 16M (kernel) | 11.25
triadalt2 64K | 1.95
triadalt2 16M | 12.44
triadalt2 64K (kernel) | 5.55
triadalt2 16M (kernel) | 10.00
triadchpl 16M | 10.24

e-kayrakli commented:
@stonea and @benharsh could I ask you both to review this PR? Of course feel free to drop comments anywhere you like, but particularly:

@stonea -- obviously this makes a lot of changes in the GPU runtime implementation and testing. I'd be happy to walk you through some of the changes as reading the diff might be difficult at times.

@benharsh -- the parts that would be more interesting to you are around new `chpl_gen_comm_get_gpu` and the `put` counterpart, and how the calls to them are generated. Particularly, changes in codegen+DR+comm macro headers in the runtime. I am not super-proud of the `get`/`get_gpu` (and same for `put`) dichotomy, nor the potential to have calls to relevant comm primitives with or without the sublocale argument. Though I can't come up with a solution that's clearly better. Let me know what you think.

stonea commented on runtime/src/chpl-gpu.c:
```diff
@@ -264,9 +311,33 @@ void* chpl_gpu_mem_alloc(size_t size, chpl_mem_descInt_t description,
   return ptr;
 }
 
+void* chpl_gpu_mem_array_alloc(size_t size, chpl_mem_descInt_t description,
```
Great to see we that this is factoring out a lot of the common code from the AND and NVIDIA implementations.

I'm wondering if this is going to break things for `CHPL_GPU=cpu`, but I see that on your paratest list in the OP.

e-kayrakli replied:
Something certainly needs adjusting with that mode. I'll get to it and not worried about it at the moment.

benharsh reviewed (commented):
I generally agree with Andy's remarks about the _gpu stuff, but alas, I don't have any great suggestions. Ideally it feels like we would be able to just hand off a locale to the runtime, and the runtime would be responsible for looking at the sublocale ID if necessary. That way we wouldn't have these differing signatures and branches in codegen. But I recognize that's a larger change than what you're going for here.

e-kayrakli commented on compiler/codegen/cg-expr.cpp:
Done.

I had to google Sisyphean. I knew the myth but not the name. Thanks!

e-kayrakli commented on compiler/codegen/cg-expr.cpp:
```diff
@@ -1023,6 +1023,23 @@ static GenRet codegenRlocale(GenRet wide)
   return ret;
 }
 
+static GenRet codegenRsubloc(GenRet wide){
```
That's exactly my understanding as well. Note the name follows from `codegenRlocale`, so not something I come up with.

e-kayrakli commented on compiler/codegen/cg-expr.cpp:
```diff
@@ -1023,6 +1023,23 @@ static GenRet codegenRlocale(GenRet wide)
+                              codegenWideThingField(wide, WIDE_GEP_LOC)),
+                          /*ln*/codegenZero(), /*fn*/codegenZero32());
+  } else {
+    USR_FATAL("--llvm-wide-opt can't be used with the GPU locale model");
```
See https://bpb-us-e1.wpmucdn.com/blogs.rice.edu/dist/1/2385/files/2013/07/Chapel_LLVM_camera_ready-q6usv4.pdf. I think it is a nice optimization that I wish to make a default one. It may also help GPUs nicely. Though it has some issues: https://github.com/Cray/chapel-private/issues/2121. We'd need to fix them first before making it work here.

e-kayrakli commented on compiler/codegen/cg-expr.cpp:
```diff
@@ -3831,13 +3855,20 @@ void codegenAssign(GenRet to_ptr, GenRet from)
-                      codegenSizeof(type),
-                      genCommID(info),
-                      info->lineno, gFilenameLookupCache[info->filename]);
+          std::vector<GenRet> args;
```
I agree, but I want to pass since also you don't seem to be pushing hard.

Handling that call to `codegenValuePtr` will make the helper unnecessarily hard to read in my view. But "wait till the third time" is a strategy I can get behind.

e-kayrakli commented on compiler/codegen/cg-expr.cpp:
```diff
@@ -5311,31 +5342,50 @@ static void codegenPutGet(CallExpr* call, GenRet &ret) {
 
+    // we actually have 5 arguments with sublocale, but than line and file info
+    // are added.
+    bool hasSubloc = (call->numActuals() == 7);
```
The type checking part can be a bit convoluted as there are many args, but I believe I can stash that away in an helper, I may give it a try based on the conversation below with Andy.

e-kayrakli replied:
https://github.com/chapel-lang/chapel/pull/22349/commits/66704e7a957dc0af431b6a45d90682b73f2cea00 checks all 5 user-given arguments. Other 2 are line/file info. Does that look good?

benharsh replied:
Looks good to me!

e-kayrakli commented on compiler/codegen/cg-expr.cpp:
```diff
@@ -5311,31 +5342,50 @@ static void codegenPutGet(CallExpr* call, GenRet &ret) {
 
+    // we actually have 5 arguments with sublocale, but than line and file info
+    // are added.
+    bool hasSubloc = (call->numActuals() == 7);
```
> I'm actually confused how this function is going to get called for the GPU case anyway. Don't we generate a call to chpl_comm_get_gpu, I don't see that in our primitives table, do we need to add it?

That would be one way of what I am doing here and I wouldn't be opposed. But there are primitives and runtime functions. For a GET, there are two runtime functions but one primitive. Which runtime call to make is decided based on the number of arguments.

I agree that the number of arguments is a shaky ground. I can take up Ben's suggestion of also checking for argument types to make things more robust. If we were to add the GPU version of the primitive instead, I'd be a bit worried of having code like:

```
if CHPL_LOCALE_MODEL=="gpu" then
  __primitive("chpl_comm_get_gpu");
else
  __primitive("chpl_comm_get");
```

So, I am for putting the ugliness in the compiler code instead of the module code. But as is (without further checks) this maybe way too ugly.

Are you game with tightening based on the argument types and providing meaningful error messages for wrong cases?

stonea replied:
I agree I don't like the idea of having these checks for `CHPL_LOCALE_MODEL` sprinkled throughout the module code. I assume there are some cases where (regardless of the state of CHPL_LOCALE_MODEL) you'd still want to just directly call `chpl_comm_get`?: otherwise couldn't we just keep what you have and move the check for `CHPL_LOCALE_MODEL==gpu` into `codegenPutGet` directly?

Anyway, assuming that wouldn't work (and even if did), I'd say at a minimum this comment needs to be updated because it's just inaccurate:

```
    // args are:
    //   localvar, locale, remote addr, get(4)==size, line, file
    //                                  get(4)==len  for array_get/put
```

But I like what Ben suggested: add some type checking checking code to identify which "overload" of the primitive is being used and add to that (in true defensive programming fashion) an assertion if we're fall into some new or invalid state.

Really, I think we should be doing this for all the primitives (it's good to be defensive and it's a lot more self-documenting than a comment that quickly gets forgotten about and falls out of date).  I don't think I'm going to be able to convince anyone (including myself) to update the existing primitives, but we can fix things as we go and a polymorphic-primitive like this one seems like a prime candidate for some checks like this.

Anyway, if it were me I'd probably timebox an hour, give it a shot, and if ended up being a hassle or the code ended up uglier, then I'd just update the comment and move on.

e-kayrakli replied:
> Really, I think we should be doing this for all the primitives (it's good to be defensive and it's a lot more self-documenting than a comment that quickly gets forgotten about and falls out of date). I don't think I'm going to be able to convince anyone (including myself) to update the existing primitives, but we can fix things as we go and a polymorphic-primitive like this one seems like a prime candidate for some checks like this.

I agree with this. I guess the main complication is that primitives could be created/consumed anytime during compilation before or after resolution. Maybe a Primitive type in the AST (as opposed to an enum and a flag on CallExpr) can provide a better infra for that support. dyno might be doing something of that sort.

I am currently considering sprinkling assertions in this function. i.e. whenever we pick a new `curArg` we check its type. I'll see what that turns out to be like.

e-kayrakli replied:
https://github.com/chapel-lang/chapel/pull/22349/commits/66704e7a957dc0af431b6a45d90682b73f2cea00 checks all 5 user-given arguments. Other 2 are line/file info. Does that look good?

stonea replied:
> I am currently considering sprinkling assertions in this function. i.e. whenever we pick a new curArg we check its type. I'll see what that turns out to be like.

Ok, thanks for doing that, it feels more safe to me.

I think this comment is still out of date, please update it if you agree:

```
    // args are:
    //   localvar, locale, remote addr, get(4)==size, line, file
    //                                  get(4)==len  for array_get/put
```

e-kayrakli replied:
2be1aa3f29

e-kayrakli commented on compiler/codegen/cg-expr.cpp:
```diff
@@ -5345,38 +5395,78 @@ static void codegenPutGet(CallExpr* call, GenRet &ret) {
+        // runtime call
+        GenRet subloc;
+
+        if (curArg->isRefOrWideRef()) {
```
Good point. It appears elsewhere in the existing code, too. https://github.com/chapel-lang/chapel/pull/22349/commits/de1e71f2cb8d91ce920a0cca44ef519c89da9e5b is a nice refactor.

e-kayrakli commented on runtime/include/chpl-comm-compiler-macros.h:
```diff
@@ -61,6 +62,28 @@ void chpl_gen_comm_get(void *addr, c_nodeid_t node, void* raddr,
 }
+#ifdef HAS_GPU_LOCALE
+static inline
+void chpl_gen_comm_get_gpu(void *addr, c_nodeid_t src_node,
```
I am not opposed to your suggestion. I have a slight preference for keeping `chpl_gen_comm_get_` as the "prefix". How do you like `chpl_gen_comm_get_from_subloc` and `chpl_gen_comm_put_to_subloc` ? I can also use `gpu` instead of `subloc`. Though technically this function is not required to involve gpu-based allocations. So "aware" is a more precise word for sure. However, I can hide behind the fact that `c_sublocid_any` or `c_sublocid_none` are also a subloc :)

stonea replied:
> How do you like chpl_gen_comm_get_from_subloc and chpl_gen_comm_put_to_subloc ?

I actually prefer that on the basis that it future- proofs things for the theoretical future where we have other kinds of sublocs.

e-kayrakli replied:
https://github.com/chapel-lang/chapel/pull/22349/commits/05c8e80e50ee5589e49fa590a1c07a14365f0554

e-kayrakli commented on runtime/include/chpl-comm-compiler-macros.h:
```diff
@@ -61,6 +62,28 @@ void chpl_gen_comm_get(void *addr, c_nodeid_t node, void* raddr,
+  c_sublocid_t dst_subloc = chpl_task_getRequestedSubloc();
+
+  if (chpl_nodeID == src_node) {
+    chpl_gpu_memcpy(dst_subloc, addr, src_subloc, raddr, size, commID, ln, fn);
```
Yeah, I am a bit more reluctant to a rename here.

e-kayrakli replied:
Do you want to argue for a change?

stonea replied:
Well it's one of those thing I'm not particularly happy with but I don't have any constructive suggestions so I'm willing to shrug it off after complaining a bit (I dislike my suggestion of `chpl_gpu_aware_memcpy` too since it would be mixing the "name" of the function with the convention of having `chpl_gpu` serve as our "poor man's module namespacing" identifier (and `chpl_gpu_gpu_aware_memcpy` is just.. well. no.).

But to voice my complaint (so I can put it rest): if I was new to the project and you told me there was two functions: `chpl_gpu_memcpy` and `chpl_memcopy` and asked me to guess what these two do.  I would guess that `chpl_memcopy` would be the more "abstract" version that would figure out what kind of copy you need to do and dispatch to, and `chpl_gpu_memcpy` would be that more "concrete" function that gets called if/when a GPU is involved (rather than the other way around).

e-kayrakli replied:
> But to voice my complaint (so I can put it rest): if I was new to the project and you told me there was two functions: chpl_gpu_memcpy and chpl_memcopy and asked me to guess what these two do. I would guess that chpl_memcopy would be the more "abstract" version that would figure out what kind of copy you need to do and dispatch to, and chpl_gpu_memcpy would be that more "concrete" function that gets called if/when a GPU is involved (rather than the other way around).

Totally reasonable.

Some alternatives as food for thought:

- `chpl_gpu_comm`
- `chpl_gpu_comm_get`/`chpl_gpu_comm_put`

(though I imagine we'll probably need something more `comm`-y down the road)

- `chpl_gpu_data_copy` 

(tries too hard to avoid just calling it `memcpy` imo)

- `chpl_memcpy_subloc`

(is interesting, but doesn't have `chpl_gpu` prefix. `chpl_gpu_memcpy_subloc` is too tautological)

I am still going to stick with what I have unless you want to champion either one of these or something else.

e-kayrakli replied:
I guess in the future, we may have `chpl_memcpy_subloc` elsewhere in the runtime. That may end up calling some `chpl_gpu` function..

stonea replied:
> I am still going to stick with what I have unless you want to champion either one of these or something else.

Nope, I'm fine, just keep these ideas in the back-of-your-head incase we have another opportunity to rename/refactor it.

e-kayrakli commented on runtime/include/gpu/amd/chpl-gpu-gen-includes.h:
```diff
@@ -77,6 +77,10 @@ __device__ static inline c_nodeid_t chpl_rt_nodeFromLocaleID(chpl_localeID_t loc
   return loc.node;
 }
 
+__device__ static inline c_nodeid_t chpl_rt_sublocFromLocaleID(chpl_localeID_t loc) {
```
https://github.com/Cray/chapel-private/issues/2416 also has some info/discussion about creating a separate runtime binary for device functions. Which we'd eventually need, I think.

I agree about creating a common header for vendors. I'll look into it.

e-kayrakli replied:
https://github.com/chapel-lang/chapel/pull/22349/commits/2440ac3d81b51e5668c9010553a3e4da6d39bb5a was a fun refactor. I think we could go further with atomics, but I didn't want to reverse engineer the macros there.

stonea replied:
Thanks! It's always a good day when your lines-of-code count goes down.

e-kayrakli commented on modules/internal/DefaultRectangular.chpl:
```diff
@@ -2055,9 +2057,13 @@ module DefaultRectangular {
     const Aidx = A.getDataIndex(Alo);
     const Adata = _ddata_shift(A.eltType, A.theData, Aidx);
     const Alocid = Adata.locale.id;
+    const Asublocid = if CHPL_LOCALE_MODEL != "gpu" then c_sublocid_any else
```
I agree. But somehow I have a recollection where we were trying to avoid `_none` for some reason. I'll grep around a bit more, and may take your suggestion based on what I found.

e-kayrakli replied:
Soo, for a task, `_any` means the task didn't target any sublocale. `_none` means it specifically targeted the host. See:

```chpl
extern proc chpl_task_getRequestedSubloc(): int;

writeln(chpl_task_getRequestedSubloc());  // -2


const host = here;

on here.gpus[0] {
  writeln(chpl_task_getRequestedSubloc());  // 0

  on host {
    writeln(chpl_task_getRequestedSubloc());  // -1
  }
}

on here {
  writeln(chpl_task_getRequestedSubloc());  // -1
}
```

For the time being `_any` feels the right constant here as `_none` is more strict and I don't know the implications of that down the road. Today, though it shouldn't matter with this PR as the runtime is checking for `<0`.

I think we need to think a bit more about these constants at some point. Maybe renaming `_none` to `_cpu` is an improvement?

e-kayrakli replied:
To be clear, I don't intend to make that final suggestion happen in that PR. I like the idea but dunno the general implications on runtime implementation's clarity.

stonea replied:
> Soo, for a task, _any means the task didn't target any sublocale. _none means it specifically targeted the host. See:

This is my understanding too.

Right, so I guess this LOC is saying "if we're not using the gpu locale model don't make any assumptions about what sublocale (which may be, perhaps, NUMA domains) that this data is on.  Sounds right.

Anyway, you can consider this item resolved.

e-kayrakli commented on runtime/src/chpl-gpu.c:
```diff
@@ -78,32 +78,10 @@ void chpl_gpu_init(void) {
 }
 
 void chpl_gpu_support_module_finished_initializing(void) {
-  // The standard module has some memory that we allocate when we  are "on" a
```
> I'm not sure I get (1)

IIRC, there was a case where we didn't get the right context when we made a call to a `cuMemcpy` family function and it was part of the reason why we needed that ugly workaround. That's a bit handwave-y but this is an old branch now :)

> o we want the locale instance stored on the GPU?

That's an interesting question. I think today with aod, the instance sits in the GPU memory, but CPU can manipulate it. However, `on here.gpus[n]` doesn't GET the instance to the host. At least it shouldn't. `here.gpus[n]` by itself should be a wide reference that has the locale ID. As long as you don't put that in a `var` or `const` it should be safe. I'll check to be aware of what's going on. If problematic, I'll probably want to fix that in a subsequent PR as the solution is not very clear to me.

e-kayrakli replied:
Looks like we're good and actually better than I thought. None of the following cases generate device to/from host comm:

```chpl
use GpuDiagnostics;


{ // basic case
  startGpuDiagnostics();
  on here.gpus[0] {
    stopGpuDiagnostics();
    writeln(getGpuDiagnostics());

    doStuff();
  }
}

{ // through ref
  startGpuDiagnostics();
  ref gpu = here.gpus[0];
  on gpu {
    stopGpuDiagnostics();
    writeln(getGpuDiagnostics());

    doStuff();
  }
}

{ // through var
  startGpuDiagnostics();
  var gpu = here.gpus[0];
  on gpu {
    stopGpuDiagnostics();
    writeln(getGpuDiagnostics());

    doStuff();
  }
}

{ // through var + field access
  startGpuDiagnostics();
  var gpu = here.gpus[0];
  var gpuId = gpu.id;
  on gpu {
    stopGpuDiagnostics();
    writeln(getGpuDiagnostics());

    doStuff();
  }

  writeln(gpuId);
}


proc doStuff() {
  use GPU;

  var A: [1..10] int;
  foreach a in A do {
    assertOnGpu();
  }
}
```

e-kayrakli commented on runtime/src/gpu/nvidia/gpu-nvidia.c:
```diff
@@ -64,8 +64,8 @@ static bool chpl_gpu_has_context(void) {
 
-static void chpl_gpu_switch_context(int deviceId) {
-  CUcontext next_context = chpl_gpu_primary_ctx[deviceId];
+static void switch_context(int dev_id) {
```
Yeah. If it wasn't clear, I like namespacing things :)

I am also open to prefix names with something to emphasize that they are `static`. Seeing a call to this function, I went "oh no, are we exposing this in the top layer?", luckily that's not the case. That's why I changed the name.

stonea replied:
Ok cool. That makes sense, I just want to know what the convention is (even if it's just internal to the gpu runtime libs) so I can follow it going forward.

e-kayrakli commented on runtime/src/gpu/nvidia/gpu-nvidia.c:
```diff
@@ -64,8 +64,8 @@ static bool chpl_gpu_has_context(void) {
 
-static void chpl_gpu_switch_context(int deviceId) {
-  CUcontext next_context = chpl_gpu_primary_ctx[deviceId];
+static void switch_context(int dev_id) {
```
> Ok, I suspect the answer is "it's chaos, we're not consistent, and there's bigger fish to fry than naming conventions", but it would be nice if I knew what I should be trying to do going forward so I could avoid adding to the chaos.

If me making the call on this helps, let's go with `snake_case` in the GPU code in the runtime. To me the only exception should be if we're naming a runtime wrapper of a an exported Chapel proc that has camelCase name. e.g. `chpl_gpu_fooBar` is ok if it wraps around `export proc fooBar`.

I know the runtime is chaos with capitalization, but I think snake_case is more common. It certainly is the case for the GPU code. Of course you can say that "you showing your own code as precedence is not fair" :)

e-kayrakli commented on runtime/include/chpl-comm-compiler-macros.h:
```diff
@@ -102,6 +125,30 @@ void chpl_gen_comm_put(void* addr, c_nodeid_t node, void* raddr,
 }
+#ifdef HAS_GPU_LOCALE
+static inline
+void chpl_gen_comm_put_gpu(void* addr,
```
https://github.com/chapel-lang/chapel/pull/22349/commits/05c8e80e50ee5589e49fa590a1c07a14365f0554

e-kayrakli commented on runtime/include/gpu/amd/chpl-gpu-gen-includes.h:
```diff
@@ -204,6 +208,16 @@ GPU_3OP_ATOMIC(unsigned int,           chpl_gpu_atomic_CAS_uint,      atomicCAS)
 // [*] GPU_3OP_ATOMIC(unsigned short int,     chpl_gpu_atomic_CAS_ushort, atomicCAS);
 
+__device__ static inline
+void chpl_gen_comm_get_gpu(void *addr, c_nodeid_t src_node,
```
https://github.com/chapel-lang/chapel/pull/22349/commits/e1abe72037ea9ae9ac1a1bf4d598f354c43c9685

e-kayrakli commented on test/gpu/native/GpuTestCommon.chpl:
```diff
@@ -0,0 +1,38 @@
+
+    const expected = if isUm then um else aod;
+
+    const diags = getGpuDiagnostics()[0];
```
I agree. We also need support for `.*DiagnosticsHere()` family of functions while we are on the subject of GPU diagnostics in a multilocale config.

e-kayrakli commented on test/gpu/native/GpuTestCommon.chpl:
```diff
@@ -0,0 +1,38 @@
+
+  config const printGpuDiags = false;
+
+  proc verifyDiags(um, aod, h2d=-1, d2h=-1, d2d=-1) {
```
cd22d47558 does a rename similar to what you suggest.

I think you said something about this already somewhere, but having to pass `-M` everytime you `chpl` a test (instead of `start_test`) could get annoying real fast. We can consider adding this function into `GpuDiagnostics` directly.

e-kayrakli commented on test/gpu/native/distArray/blockOutsideOnWorkaround.skipif:
```diff
@@ -0,0 +1,6 @@
+# with aod, the access to arr's elements are handled with a get. However, it
```
Working on this a bit made me a little bit pessimistic about our distributed array story. I don't think this can get fixed without us knowing it. So, I'm ok with `skipif`. We'll need a good amount of effort to get things working here.

e-kayrakli commented on test/gpu/native/kernelFnCalls/callTrivialFn.good:
```diff
@@ -1 +1 @@
-(kernel_launch = 1)
```
bd3a01a4af

e-kayrakli commented on test/gpu/native/kernelFnCalls/fnWithForall.chpl:
```diff
@@ -23,7 +25,11 @@ on here.gpus[0] {
   }
 
-  writeln(A);
+  on host { // copy data back to avoid interleaving output with verbose GPU
```
https://github.com/chapel-lang/chapel/issues/21754

e-kayrakli commented on test/gpu/native/streamPrototype/COMPOPTS:
```diff
@@ -1 +1 @@
---no-checks
```
(I commented somewhere above and I am lazy to find) What do you think about just putting this validation into `GpuDiagnostics` in a separate PR?

e-kayrakli replied:
I will move the validation to `GpuDiagnostics` soon after this PR. I'll probably name it `assertGpuDiags`. I might align the argument names to the field names of the GPU diagnostics.

e-kayrakli commented on test/gpu/native/studies/transpose/transpose.chpl:
```diff
@@ -115,45 +116,55 @@ inline proc transposeLowLevel(original, output) {
+
+var elapsed = timer.elapsed() / numTrials;
+
+var sizeInBytes = originalHost.size * numBytes(dataType);
```
Based on my desire to move the validation function into `GpuDiagnostics` and avoiding `-M` in compopts as much as possible. I am leaning towards keeping this in at the moment.

e-kayrakli commented on test/gpu/native/useAssociativeDomain.chpl:
```diff
@@ -17,5 +19,4 @@ on here.gpus[0] {
 stopGpuDiagnostics();
-writeln("GPU diagnostics:");
-writeln(getGpuDiagnostics());
+verifyDiags(um=0, aod=1); // we are currently skipping with aod
```
Associative domains barely work with GPUs. At some point where the rudimentary support was added, I believe the main motivation was just to make sure that they compiled and run correctly on CPUs. This test's main goal is to make sure that we can compile this code.

If you look at the code (that's hidden in the diff), the `foreach` doesn't result in a kernel launch. The only launch that we get with aod is for array initialization.

e-kayrakli commented on test/gpu/native/studies/shoc/sort.chpl:
Yeah, the main tendency that we should all have is to have smaller `on` statements as they know may incur "communication" costs by unnecessarily calling get/put. ie. having a large `on` statement, and within doing your validation on host by using `for` is a nice trick, but terrible for performance. Instead the data should be bulk copied and validated inside something like `on host`.

I am planning to make some notes about that and share it with the GPU team, because I don't want us to have blanket `on` statements anymore. Probably that discussion will have the final question "are you OK with making this mode the default?".

(Not blaming anyone. We all need to adjust, definitely including me.)

e-kayrakli commented:
Thanks both for reviews. I did my first round addressing your concerns and I think this is ready for another review.

I responded to your comments with commits that fixed the issue. For some, I clicked "resolve" as I think the request was clear. Some I left unresolved as I thought it'd be great if you can check my solution. Feel free to unresolve/resolve conversations as you see fit. The "unresolved" conversations are my todo items.

@stonea -- I am aware of test unit changes you asked. I will get to them, but they are more on the trivial side and I am about to time out for today :)

benharsh reviewed (approved):
My items were resolved, thanks!

stonea reviewed (approved):
The only outstanding issue I see are updating:
- updating the comment on `codegenPutGet` (assuming I'm not misunderstanding what this comment says, please do this)
- doing something about the `-M/..` paths so we have the common module in path (please either do this or make it a short-term followup task)

For the second one, in one of your comments you mentioned an idea of having your new `verifyGpuDiags` function in `GpuDiagonstics` directly (and despite my "long live GpuTestCommon.chpl" comment) I really like this idea: (for us) it's more convenient and avoids the need to do anything special in the `.compopt` file completely.

Do you think you could get this done before merging the PR?

If not could I ask you to add it as a followup item and it to your current sprint?

This is the sort of thing that's probably trivial to fix while you still have the context, but will be annoying if/when we do want to rearrange things and find out that we've hard-coded some assumptions about relative paths into our tests. Plus, leaving it "as is", basically guarantees that others are going to pick up on it and continue adding these `-M/..` flags in new tests).

e-kayrakli commented:
> updating the comment on codegenPutGet (assuming I'm not misunderstanding what this comment says, please do this)


Will do. (You're not misunderstanding at all)

> doing something about the -M/.. paths so we have the common module in path (please either do this or make it a short-term followup task)

I will do that as a subsequent PR. The longer I hold this PR, the more chance I'll have to deal with conflicts it looks like. I do agree that it should happen sooner rather than later.

e-kayrakli commented on test/gpu/native/studies/transpose/transpose.chpl:
```diff
@@ -115,45 +116,55 @@ inline proc transposeLowLevel(original, output) {
+var gbPerSec = sizeInGb / elapsed;
+if perftest {
+  writeln("Wall clock time (s): ", elapsed);
+  writeln("Performance (GB/s): ", gbPerSec);
```
d7368097b4

will do that in this test.

There are some other tests (particularly SHOC) that seemingly only reports GB/s. Though technically it is also reporting GiB/s (divides by 1e9). But that's consistent with the reference version. So, even with a very small sample size we bump into a benchmark that doesn't conform to the standard :) We can add a `config const` there, but I'd prefer the default to be consistent with the reference. In which case, I am not sure how to setup that `config const`.

We've definitely been bitten by this before, but it is a messy subject not just for us it seems..

e-kayrakli commented:
Under https://github.com/Cray/chapel-private/issues/5135 I have a design proposal for `assertOnGpu`. I will see that it moves forward quickly and the tests are cleaned up.
